tests/test_session_0.py

```diff
@@ -0,0 +1,185 @@
+
+    except AssertionError:
+        import traceback
+        error_line = traceback.format_exc().splitlines()[-2].replace('[obj]', '['+repr(obj)+']')
+        print("Error while testing (byref=%s):" % byref, error_line, sep="\n", file=sys.stderr)
+        raise
+
+
+if __name__ == '__main__':
+
+    # Test dump_session() and load_session().
+    for byref in (False, True):
+        if byref:
+            # Test unpickleable imported object in main.
+            try:
+                from ctypes import pythonapi
+            except ImportError:
+                pass
+
+        #print(sorted(set(sys.modules.keys()) - original_modules))
+        dill._test_file = dill._dill.StringIO()
+        try:
+            # For the following test files.
+            dill.dump_session('session-byref-%s.pkl' % byref, byref=byref)
+
+            dill.dump_session(dill._test_file, byref=byref)
+            dump = dill._test_file.getvalue()
+            dill._test_file.close()
+
+            import __main__
+            copy_dict = __main__.__dict__.copy()
+            copy_modules = sys.modules.copy()
+            del copy_dict['dump']
+            del copy_dict['__main__']
+            for name in copy_dict.keys():
+                if name not in original_objects:
+                    del __main__.__dict__[name]
+            for module in list(sys.modules.keys()):
+                if module not in original_modules:
+                    del sys.modules[module]
+
+            dill._test_file = dill._dill.StringIO(dump)
+            dill.load_session(dill._test_file)
+            #print(sorted(set(sys.modules.keys()) - original_modules))
+        finally:
+            dill._test_file.close()
+
+        test_objects(__main__, copy_dict, byref)
+        __main__.__dict__.update(copy_dict)
+        sys.modules.update(copy_modules)
+        del __main__, copy_dict, copy_modules, dump
+
+
+    # This is for code coverage, tests the use case of dump_session(byref=True)
+    # without imported objects in the namespace. It's a contrived example because
+    # even dill can't be in it.
+    from types import ModuleType
+    main = ModuleType('test_module')
+    main.x = 42
+
+    _main = dill._dill._stash_modules(main)
+    if _main is not main:
+        print("There are objects to save by referenece that shouldn't be:",
+              _main.__dill_imported, _main.__dill_imported_as, _main.__dill_imported_top_level,
+              file=sys.stderr)
+
+    test_file = dill._dill.StringIO()
+    try:
+        dill.dump_session(test_file, main=main, byref=True)
+        dump = test_file.getvalue()
+        test_file.close()
+
+        # This should work after fixing https://github.com/uqfoundation/dill/issues/462
+        test_file = dill._dill.StringIO(dump)
+        dill.load_session(test_file)
+    finally:
+        test_file.close()
+
+    assert x == 42
+
+
+    # Dump session for module that is not __main__:
+    import test_classdef as module
```

Comment:
need to clean up the files left on disk (i.e. *pyc or in the cache).  See `test_moduledict_where_not_main` in `test_nested.py`.

Reply:
solved